src/ncp/hdlc.hpp

```diff
@@ -51,107 +54,390 @@ namespace ot {
+     * pointer to buffer (from `GetFrame()` or `ReadSavedFrame()`) should be considered invalid after calling this
+     * method.
+     *
+     */
+    void ClearReadFrames(void)
+    {
+        uint16_t readLen = static_cast<uint16_t>(mSavedFrameStart - mBuffer);
+
+        if (readLen > 0)
+        {
+            memmove(mBuffer, mSavedFrameStart, static_cast<uint16_t>(mWritePointer - mSavedFrameStart));
+            mWritePointer -= readLen;
+            mWriteFrameStart -= readLen;
+            mSavedFrameStart -= readLen;
+            mRemainingLength += readLen;
+        }
+    }
+
+private:
+    /*
+     * The diagram below illustrates how the frames are saved in the buffer.
+     *
+     * Each saved frame contains a header which is 2 bytes long and specifies the frame length (the length does not
+     * include the header itself). The frame length is stored in header bytes as a `uint16_t` value using little-endian
+     * encoding.
+     *
+     * The digram shows `mBuffer` and different pointers to the buffer. It represent buffer state when there are two
+     * saved frames in buffer and the first saved frame was already (`mSavedFrameStart` is pointing to header of second
+     * frame).
+     *
+     *          Saved frame #1           Saved frame #2       Current frame being written
+     *   /                        \ /                      \ /                           \
+     *   +-----------+-------------+-----------+------------+---------+--------------------------------------------+
+     *   | header #1 |   ...       | header #2 |  ...       | header  |  ...             | ...                     |
+     *   +-----------+-------------+-----------+------------+---------+--------------------------------------------+
+     *   ^                         ^                        ^                            ^\                       /^
+     *   |                         |                        |                            |   mRemainingLength      |
+     *  mBuffer[0]                mSavedFrameStart          mWriteFrameStart             |                         |
+     *                                                                                   |              mBuffer[kSize]
+     *                                                                                 mWritePointer
+     *
+     */
+
+    enum
+    {
+        kHeaderSize = sizeof(uint16_t),
     };
 
+    uint8_t  mBuffer[kSize];
+    uint8_t *mSavedFrameStart; // Pointer to start of next saved frame (for `ReadSavedFrame()`).
+    uint8_t *mWriteFrameStart; // Pointer to start of current frame being written.
+};
+
+/**
+ * This class implements the HDLC-lite encoder.
+ *
+ */
+class Encoder
+{
+public:
     /**
      * This constructor initializes the object.
      *
+     * @param[in] aWritePointer   The `FrameWritePointer` used by `Encoder` to write the encoded frames.
+     *
      */
-    Encoder(void);
+    Encoder(FrameWritePointer &aWritePointer);
```

Comment:
Declare as `explicit`?

Reply:
All changed in the new pushed commit. Thanks.